scripts/subset_ntopo.py

```diff
@@ -0,0 +1,219 @@
+                'reachCount',
+                'upReachStart',
+                'upReachCount',
+                'upHruStart',
+                'upHruCount']
+# Variables with sHRU dimension
+hru_var_name = ['hruIndex',
+                'hru_id',
+                'hru_lon',
+                'hru_lat',
+                'hru_elev',
+                'hru_area',
+                'hru_weight']
+# Variables with sUps dimension
+ups_var_name = ['upReachIndex',
+                'upReachID']
+# Variables with sAll dimension
+allups_var_name = ['reachList',
+                   'upReachTotalLength']
+
+
+def read_ntopo(nc_in):
+    ntopo_data = xr.open_dataset(nc_in)
+    return ntopo_data 
+
+
+def write_ntopo(ntopo_data, nc_out):
+    ntopo_data.to_netcdf(nc_out, format='NETCDF4')
+    return None
+
+def subset_ntopo(seg_id, in_data):
+    """Find upstream segs of <seg_id> in network data <in_data>
+       extract inforation of upstream segs and return them in <out_data> 
+    """
+    print('Outlet segment = %s'%seg_id)
+
+    # Identify index of the desired stream segment from reachID vector
+    reach_id_array = in_data['reachID'].values     # reach ID
+    idx = np.where(reach_id_array == int(seg_id))
+    if np.size(idx) != 1:
+        raise Exception('unable to find desired stream segment')
+
+    # Split into separate datasets based on a dimension
+    seg_data = in_data
+    hru_data = in_data
+    ups_data = in_data
+    for varname in in_data.data_vars:
+        if in_data[varname].dims[0] != 'sSeg':
+            seg_data = seg_data.drop(varname)
+        if in_data[varname].dims[0] != 'sHRU':
+            hru_data = hru_data.drop(varname)
+        if in_data[varname].dims[0] != 'sUps':
+            ups_data = ups_data.drop(varname)
+
+    idx_rch_start = in_data['reachStart'].values[idx]-1
+    n_rch_count = in_data['reachCount'].values[idx]
+    print 'Number of upstream segment from outlet segment (nRchCount) = %d'%n_rch_count
+                         
+    # Read reach list of index from global segments (all the upstream reachs for each segment) 
+    idxs_uprch_list = in_data['reachList'].values[idx_rch_start[0]:idx_rch_start[0]+n_rch_count[0]]
+
+    # Reach upstream segment and associated HRU infor from non-ragged vector 
+    seg_sub_data = seg_data.isel(sSeg = idxs_uprch_list-1)
+    print 'upstream seg id list = ',seg_sub_data['reachID'].values
+
+    # Redo index for 'reachIndx' as local segment list
+    seg_sub_data['reachIndex'] = xr.DataArray(np.arange(1,n_rch_count+1), dims=['sSeg'])
+
+    # Re-compute downstream segment index as local segment list
+    down_idx_list = np.full(n_rch_count, FILL_VALUE, dtype=np.int32)
+    for ii, down_id in enumerate(seg_sub_data['downReachID'].values):
+        down_idx = np.where(seg_sub_data['reachID'].values == down_id)[0]
+        if np.size(down_idx) == 1:
+            down_idx_list[ii] = down_idx+1
+    seg_sub_data['downReachIndex'] = xr.DataArray(down_idx_list, dims=['sSeg'])
```

Comment:
So 
`reachID` is a list of segment ID
 `downReachIndex` is a list of indices of `reachID` but the index in `downReachIndex` point to the immediate downstream segment. downstream segment ID exist somewhere in `reachID`. 

both has the same dimension size.  

when I subset river network data, I need to redo all the indices because they are based on the parent network data.  

So here, I am trying to redo index for down stream segment by using list of reach ID and downstream reach ID for each segment.

I am not sure how i can use `where` function to get all the indices for each array elements without looping.